### ISSUE TITLE
Challenges Around Upgrading PostHog With Async Migrations

### ISSUE THREAD

**Issue by yakkomajuri** [PostHog/posthog#8691]
## Overview

We've introduced a system to run async migrations such that we can get people from outdated database schemas to the updated schemas we will support going forward. Unlike traditional migrations, however, there is no expectation that these can be run in a synchronous manner, as they require data migrations that can be immediate for small users but might take weeks for large users. Async migrations specify no upper bound on how long they may take.

Another constraint is that our developers need to have certainty that as of some specific point they may begin writing code with the expectation that the migration has been completed. Otherwise there's no point in the migrations existing. 

As such, we're left with the problem of upgrades. How can we ensure the best possible upgrade experience given these constraints?

## Problems

As of right now, async migrations specify an upper bound in the form of a PostHog version, after which they are a hard requirement.

For example, for the events migration (0002), this is 1.34.0. If someone tries to boot up PostHog 1.34.0 without having completed that migration, booting it up will fail. This guarantees that developers writing 1.34.0 code can rely on the schema being up to date.

The issue of fresh deploys is already solved. Users deploying a new instance will already have the newest schema, so we will skip the async migrations. This is noted in our [Handbook's guide to writing an async migration](https://posthog.com/handbook/engineering/async-migrations#scope-and-limitations#scope-and-limitations).

Upgrading from a version where the migration already exists is also a problem that's solved. If your upgrade fails, it will rollback and you will stay in the same PostHog version until the migration completes, after which you can then upgrade with no problems.

The biggest problem is upgrading from a version before the existence of the migration to a version where it is required. e.g. moving from 1.30.0 to 1.34.0.

We **cannot** run the migration synchronously during the migrate step in this case, because for a large customer it could take e.g. a week, and thus the whole point of async migrations would be defeated. 

Thus, the current approach is that the user would need to upgrade to 1.32.0, run the migration, and then move to 1.34.0+.

One of the problems here, however, is that, while we will throw an error denoting that PostHog will not boot because of the migration, `helm upgrade` will not feed this back to the user. 

So, what should we do?

### Implementation questions

1. Can we issue a timeout-less `helm upgrade` that can run in the background while the old version keeps running? If this is the case, we can force a synchronous run of the migration causing PostHog to only boot up the newest version when that's done. (@guidoiaquinti thoughts?)
2. Can we boot up a minimal version of PostHog where we still ingest events and allows for async migrations management on such an upgrade?
3. What else can we do?

### Dev experience actionables

- [ ] Give the boot error more specific instructions on what to do. Instead of just denoting why it failed, instruct the user to upgrade to X and then update again
- [ ] Ensure release notes have appropriate documentation on upgrading 
- [ ] Bring back automatic runs of migrations so that we immediately kick them off in the background when a user upgrades to a version above the minimum to run the migration? (cc @tiina303) 

cc @macobo @guidoiaquinti @tiina303

**Comment by macobo** [comment]
Related aside: Upgrading clickhouse _requires_ that the events async migration has been run (or it's a fresh install), meaning that if we change the default version it also creates a breaking change for our users.

Nothing to speak of hobby install - not sure how that works with anything here :sweat_smile: 

> Can we issue a timeout-less helm upgrade that can run in the background while the old version keeps running?

Helm install always has a timeout (we recommend 30m) and since it's running/blocking from the users CLI this wouldn't work. 

Currently how upgrades would fail is due to the timeout I assume. Given we have ways to verify everything is bootable I wonder if we can change things so we can provide faster feedback (e.g. some sort of init container)?

As for how to avoid downtime if someone overeagerly updates - not sure if we already have this w/ blue-green deploys, @guidoiaquinti has the context here.

**Comment by guidoiaquinti** [comment]
With the latest version of the chart, a new deployment will not be considered healthy by the k8s control plane till [this](https://github.com/PostHog/charts-clickhouse/blob/29198e2f05e087e82e102dc654d0c7012a7e28fb/charts/posthog/templates/_snippet-initContainers-wait-for-migrations.tpl#L9) returns 0. My suggestion to @tiina303 was to change that logic to execute the new `bin/check-migration` so that a Helm upgrade containing an unsupported release will never succeed (as the application pods will never reach readiness status).

**Comment by macobo** [comment]
> so that a Helm upgrade containing an unsupported release will never succeed (as the application pods will never reach readiness status).

Let's dig into this a bit more. You say the upgrade will fail - will it time out or will it exit early?

Let's focus on the chart operator standpoint - how can we ensure users have a smooth upgrade path between releases? How will users find out what they need to do next if they try to upgrade from (say) 1.30 to 1.35 in one go?

**Comment by guidoiaquinti** [comment]
> You say the upgrade will fail - will it time out or will it exit early?

I think it's up to us to decide about how to deal with the error. 

Currently the init container implementation will loop till a timeout is reached. We can also decide to put this in the migrate job and let it hard fail.

**Comment by macobo** [comment]
> I think it's up to us to decide about how to deal with the error.

Let's figure out what the desired behavior is, what the current behavior is, spec it out here. That's kind of what this issue is about. :)

**Comment by tiina303** [comment]
@yakkomajuri  and I discussed this and the proposal is:
1. First thing we do when someone runs a `helm upgrade` we check async migrations (before we run django migrations as those could depend on the async migrations being done).
1. If that check fails we error in a way that's clear to the user and link them to the upgrade notes (technical details of how that can be done tbd, but I assume it's doable)
1. users need to upgrade in steps if they are using really old versions (they really shouldn't be using really old versions, so if they do it's acceptable for the upgrade to be more complex)

**Comment by macobo** [comment]
>  (technical details of how that can be done tbd, but I assume it's doable)

I think this _might_ be impossible as helm always will just fail with something along the lines of "hey this failed since service/pod Y didn't come up". I might be mistaken though, but I wouldn't hold my breath on being able to provide useful error messages for `helm upgrade` command output.

EDIT: quick google:
- https://stackoverflow.com/questions/63919531/is-there-any-way-to-print-custom-error-message-when-kubernetes-pre-install-job-f
- https://github.com/helm/helm/issues/3370

> users need to upgrade in steps if they are using really old versions (they really shouldn't be using really old versions, so if they do it's acceptable for the upgrade to be more complex)

Just note that this needs an accompanying docs update.

**Comment by tiina303** [comment]
> I think this might be impossible as helm always will just fail with something along the lines of "hey this failed since service/pod Y didn't come up". I might be mistaken though, but I wouldn't hold my breath on being able to provide useful error messages for helm upgrade command output.

Sure then the alternative is we document that if migration job failed how to check logs or what to do exactly (aka docs)

> Just note that this needs an accompanying docs update.

yup that's the plan

**Comment by tiina303** [comment]
I've been looking into this a bit and the simplest way is (during both install and upgrade):
1. we run all django and CH migrations
2. run async migrations --check
3. upgrade web/events/... pods to use new code

This would mean:
1. Django and CH migrations can't depend on async migrations
2. Earlier version of PostHog code needs to work on top of lates migrations (this was already a nice to have before as it would cause downtime like the incident on 18th of Feb https://docs.google.com/document/d/1pujAAxk7KVtL3q3tvAmo24PfWb2W-QCoOAnZ7EbWCgI), but now becomes more critical as async migrations aren't fast and ideally we'll be able to access the UI to kick them off, so if a migration would take down the web app with older code that wouldn't be good for the users (they either need to rollback the migration or while down run the async migrations)

If we want to make django or CH migrations depend on async migrations we need to create a more complex migration state checking and staggering. Imagine this is the dependency order - the row above must be completed first:

| id       | migration-type |
|------|-----------------|
| 0098| django                |
| 0010| ch                   |
| 0003| async                   |
| 0020| django                 |
| 0005| async                   |
| 0015| ch                      |

during install:
1. run all django and CH migrations
2. run async migrations check (likely want a better than current system for making sure completion is good, otherwise required later could fail and cause problems)

during upgrade:
1. update django migrations to 0098
2. run ch migrations to 0010
3. run async migrations check for migrations up to 0003, bail out if not completed & we don't run the django migrations that depend on these async migrations.
4. run django migrations up to 0020
5. ...

^ The way we could do this in theory is to make the `/bin/migrate` script sth like:
```python
from django.db.migrations.recorder import MigrationRecorder

migration_deps = [
'0098': 'django',
'0010': 'ch',
...
]

for id, type in migration_deps:
  if type == django:
    lm = MigrationRecorder.Migration.objects.filter(app='posthog').last()
    if lm < id:
      # run migrations up to id
  elif type == ch:
    # similar to above
  elif type == async:
    # run check up to id, if not OK then bail out and tell users they need to run the async migration & potentially what version of Posthog app they need to upgrade to.
```

This also gets complex from the chart perspective as we kind of make the assumption that chart is used with the default posthog version (e.g. I've removed envs, what bin/scripts do etc)

**Comment by tiina303** [comment]
hmm... alternatively we could just do the restrictions by posthog-version and check that first && then require using chart version X. Then we can use some validation in helm similar to the file name check perhaps, which did show-up for the user directly.
For hobby/docker directly users can validate that somehow too maybe?

**Comment by tiina303** [comment]
So the version idea is not great as depending on where you update from we might not need an async migration and this would be pessimistic and hence annoying to users. Going with the simple solution for now after discussing with @macobo that we don't currently need to dependencies between django, CH, async migrations.
